### PR TITLE
[Merged by Bors] - feat(FieldTheory): minimal polynomials determine an algebraic extension (Isaacs 1980)

### PR DESCRIPTION
**Pull request by alreadydone**
If E/F is an algebraic extension and K/F is an arbitrary extension, and if the minimal polynomial of every element of E over F has a root in K, we show there exists an F-embedding from E into K. If E/F and K/F have the same set of minimal polynomials, they are then isomorphic. Another corollary is that an algebraic extension E/F is an algebraic closure if every (monic irreducible) polynomial in F[X] has a root in E. These results are proven in FieldTheory/Isaacs.

An important definition (`IsExtendible`) and two important lemmas towards these theorems concern `IntermediateField.Lifts` and reside in FieldTheory/Extension. The lemmas say that the supremum of a chain of extendible lifts is also extendible, and that a lift with full domain exists if the bottom lift is extendible. We do slight refactoring to extract `Lifts.union`, and provide new API lemmas for Lifts.

---

Thanks to #13047 authors @AntoineChambert-Loir @bustercopley 

- [x] depends on: #18430
- [x] depends on: #18431

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Field/Subfield.lean [hunk 1]
```diff
@@ -616,14 +616,22 @@ theorem closure_eq_of_le {s : Set K} {t : Subfield K} (h₁ : s ⊆ t) (h₂ : t
...
-    exact (closure_le (t := this)).2 mem h
+      { carrier := {x | ∃ hx, p x hx}
+        mul_mem' := by rintro _ _ ⟨_, hx⟩ ⟨_, hy⟩; exact ⟨_, mul _ _ _ _ hx hy⟩
+        one_mem' := ⟨_, one⟩
+        add_mem' := by rintro _ _ ⟨_, hx⟩ ⟨_, hy⟩; exact ⟨_, add _ _ _ _ hx hy⟩
+        zero_mem' := ⟨zero_mem _, by
+          simp_rw [← @add_neg_cancel K _ 1]; exact add _ _ _ _ one (neg _ _ one)⟩
+        neg_mem' := by rintro _ ⟨_, hx⟩; exact ⟨_, neg _ _ hx⟩
+        inv_mem' := by rintro _ ⟨_, hx⟩; exact ⟨_, inv _ _ hx⟩ }
+    ((closure_le (t := this)).2 (fun x hx ↦ ⟨_, mem x hx⟩) h).2
```
Comment (alreadydone) [inline Mathlib/Algebra/Field/Subfield.lean:634]: This and the other one in Field/Adjoin are follow-ups to #17543, cc @j-loreaux

**Comment by jcommelin** [comment]
Thanks! If you split the changes to

- Mathlib/FieldTheory/IsAlgClosed/Basic.lean
- Mathlib/FieldTheory/Minpoly/Field.lean
- Mathlib/FieldTheory/PrimitiveElement.lean
- Mathlib/GroupTheory/CosetCover.lean
- Mathlib/RingTheory/Adjoin/Basic.lean

into a separate PR, then I would merge that one immediately.

I'll now look at the other files.

**Comment by jcommelin** [comment]
The files

- Mathlib/Algebra/Field/Subfield.lean
- Mathlib/FieldTheory/Adjoin.lean

also look good to me, and I would merge a PR that contains just those changes.

File: Mathlib/FieldTheory/Isaacs.lean [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+  have splits s (hs : s ∈ S) : (minpoly F s).Splits (algebraMap F K') := by
+    apply splits_of_splits_of_dvd _
+      (Finset.prod_ne_zero_iff.mpr fun _ _ ↦ minpoly.ne_zero <| (alg.isIntegral).1 _)
+      ((splits_map_iff _ _).mp <| SplittingField.splits p) (Finset.dvd_prod_of_mem _ hs)
+  let K₀ := (⊥ : IntermediateField K K').restrictScalars F
+  let FS := adjoin F (S : Set E)
+  let Ω := FS →ₐ[F] K'
+  have := finiteDimensional_adjoin (S := (S : Set E)) fun _ _ ↦ (alg.isIntegral).1 _
+  let M (ω : Ω) := Subalgebra.toSubmodule (K₀.comap ω).toSubalgebra
+  have : ⋃ ω : Ω, ↑(M ω) = @Set.univ FS :=
```
Comment (jcommelin) [inline Mathlib/FieldTheory/Isaacs.lean:51]: Does this work?
```suggestion
  have : ⋃ ω : Ω, (M ω : Set FS) = Set.univ :=
```

File: Mathlib/FieldTheory/Isaacs.lean [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    Set.eq_univ_of_forall fun ⟨α, hα⟩ ↦ Set.mem_iUnion.mpr <| by
+      have ⟨β, hβ⟩ := h α
+      let ϕ : F⟮α⟯ →ₐ[F] K' := (IsScalarTower.toAlgHom _ _ _).comp ((AdjoinRoot.liftHom _ _ hβ).comp
+        (adjoinRootEquivAdjoin F <| (alg.isIntegral).1 _).symm.toAlgHom)
+      have ⟨ω, hω⟩ := exists_algHom_adjoin_of_splits
+        (fun s hs ↦ ⟨(alg.isIntegral).1 _, splits s hs⟩) ϕ (adjoin_simple_le_iff.mpr hα)
+      refine ⟨ω, β, ((DFunLike.congr_fun hω <| AdjoinSimple.gen F α).trans ?_).symm⟩
+      rw [AlgHom.comp_apply, AlgHom.comp_apply, AlgEquiv.coe_algHom,
+        adjoinRootEquivAdjoin_symm_apply_gen, AdjoinRoot.liftHom_root]
+      rfl
```
Comment (jcommelin) [inline Mathlib/FieldTheory/Isaacs.lean:61]: I'm really happy that this works now. But I still don't understand why the timeout went away. Do we risk that it suddenly reappears? Can you please comment briefly on how you avoid it now?

File: Mathlib/FieldTheory/Isaacs.lean (hunk 2 above)
Comment (alreadydone) [reply Mathlib/FieldTheory/Isaacs.lean:51]: Yes, thanks!

**Comment by alreadydone** [comment]
Thanks! Opened #18430 #18431

**Comment by acmepjz** [comment]
Great work! Perhaps you should also change the TODO docstring in `Mathlib.FieldTheory.IsAlgClosed.Basic`.

**Comment by alreadydone** [comment]
> Great work! Perhaps you should also change the TODO docstring in `Mathlib.FieldTheory.IsAlgClosed.Basic`.

Thanks! Docstring updated. Isaacs' proof is not only more general but also less advanced than Conrad's, as it doesn't require the primitive element theorem or facts about (in)separable extensions. All it uses are Zorn's lemma, that a finite compositum of finite extensions is finite, that a finite extension only admits finitely many embeddings, and that a vector space over an infinite field isn't a finite union of proper subspaces; I'm tempted to classify it as "abstract nonsense".

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**